### PR TITLE
[E0183] Add missing error code explanation

### PR DESCRIPTION
**Pull request by Karrq**
This is the continuation of #61225 

CC  #61137 

Let me know if there's anything that needs change

### CODE REVIEW

File: src/librustc_typeck/error_codes.rs [hunk 1]
```diff
@@ -1588,6 +1588,38 @@ fn bar(foo: Foo) -> u32 {
 ```
 "##,
 
+E0183: r##"
+Fn* traits (`unboxed closures`) are experimental. 
+See [tracking issue #29625][iss29625] for the status of the feature.
+
+Functions must have exactly one (non self) argument, a tuple representing
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/error_codes.rs]: I'd put "self" between backquotes.

File: src/librustc_typeck/error_codes.rs [hunk 2]
```diff
@@ -1588,6 +1588,38 @@ fn bar(foo: Foo) -> u32 {
...
+the argument list [1]. 
+
+Here's an example of this error:
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/error_codes.rs]: Please replace with "Erroneous code example:".

File: src/librustc_typeck/error_codes.rs [hunk 3]
```diff
@@ -1588,6 +1588,38 @@ fn bar(foo: Foo) -> u32 {
...
+```compile_fail,E0183
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/error_codes.rs]: Please add an empty line before this one.

File: src/librustc_typeck/error_codes.rs [hunk 4]
```diff
@@ -1588,6 +1588,38 @@ fn bar(foo: Foo) -> u32 {
...
+extern "rust-call" fn echo(arg: (u32,)) -> (u32,) {
+    arg
+}
+```
+
+This feature can be enabled with:
+```
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/error_codes.rs]: Please add an empty line before this one.

**Review by GuillaumeGomez — CHANGES_REQUESTED** [review]
Good work, but not in the expected order. 😄 So, beyond my small fix requests, the global format of the explanation isn't "good". Steps:

1. First you should state what the error.
2. Followed then by "Erroneous code example:".
3. Then comes the erroneous code example.
4. Here goes the explanation of the error and how to solve it.
5. Here comes the fixed code example from `3.`.

Also, please remember that those long error explanations are not only rendered into html but can also be seen into terminals when running `rustc --explain [ERROR CODE]` (which explains why I asked you to put empty lines before code blocks).

File: src/librustc_typeck/error_codes.rs [hunk 5]
```diff
@@ -1588,6 +1588,41 @@ fn bar(foo: Foo) -> u32 {
...
 "##,
 
+E0183: r##"
+Fn* traits (`unboxed closures`) are experimental. 
+See [tracking issue #29625][iss29625] for the status of the feature.
+
+Functions must have exactly one (non self) argument, a tuple representing
+the argument list [1]. 
+
+Erroneous code example:
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/error_codes.rs]: 👍

**Comment by Karrq** [comment]
I tried to gather info from the surrounding errors about the format to follow but I wasn't really able. 
Thanks for the complete explanation, I'll reformat the error to adhere to your guidelines

File: src/librustc_typeck/error_codes.rs (hunk 2 above)
Comment (Karrq) [reply src/librustc_typeck/error_codes.rs]: Forgot this one. *facepalms*

File: src/librustc_typeck/error_codes.rs (hunk 1 above)
Comment (Karrq) [reply src/librustc_typeck/error_codes.rs]: Forgot to do this one with the last commit...

**Comment by Karrq** [comment]
I was wondering tho, what's the actual error? When is this error emitted?
I don't seem to be able to get the compiler to emit the error...

[Using `self`](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=6b0f532d3d8890d19cf68dc8472dbf32) doesn't emit any error, and actually works... bug?
[Not providing the tuple argument](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=b94bb91198bbb2557e8be93e9e0cd3ea) ICEs on `cargo run`.
[Not adding the feature gate](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=3e0fdc9fe1d3e8ea3da2c1e8649e8136) emits E0658.

**Comment by GuillaumeGomez** [comment]
> I was wondering tho, what's the actual error? When is this error emitted?
> I don't seem to be able to get the compiler to emit the error...

@Karrq That's part of why we ask to show the error. It also helps to check that the error is still emitted. It's however a bit surprising that the error doesn't show up... I can only recommend you to take a look at how it's emitted in the compiler (generally it's not very complicated). Look for the error code directly.

If you still can't figure out, maybe you can give a try to another error code?

**Comment by Karrq** [comment]
@GuillaumeGomez I was able to figure out when the error is *supposed* to be emitted: `librustc_typeck::coherence::enforce_trait_manually_implementable`, and it should be emitted when the user tries to implement any of the `Fn*` traits manually (line 78 of mod.rs). [I tried doing that](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=6c2e03be6e64b9f7c0036b6c318c26f3) but I wasn't able to get it to produce the error, again.

Upon digging more (searching `rust-call ABI is subject to change`) I was able to find some tests run for this feature. One of these files does the test to check what errors are emitted when trying to manually implement the `Fn*` traits (`src/test/compile-fail/feature-gate-unboxed-closures-manual-impls.rs`). As can be seen in the file there's a note that specifies that the error never triggers because other errors are raised before that. That note has been there since commit [09783d1dab](https://github.com/rust-lang/rust/commit/09783d1dab5817b9c0202ce2fa2b0e5e78e79d44), dated Jan 12 2015. 

I believe there's no way from us to trigger this specific error, yet. Nevertheless, now that I know what's supposed to emit this error I'll update the description accordingly in the next commit.

**Comment by Karrq** [comment]
According to the [unstable book](https://doc.rust-lang.org/unstable-book/language-features/unboxed-closures.html), `unboxed_closures` should be used to enable the user to use the `"rust-call"` ABI, but the compiler suggests to enable `fn_traits` [instead](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c11ae01fb6db5a1462be992174ad1fc0).

From what I gathered:
1. E0183 should tell the user to enable `fn_traits` since it's emitted when checking for the `Fn*` traits. 
2. When trying to use the "rust-call" ABI, the compiler should suggest to enable `unboxed_closures`. I wasn't enable to find where this error is emitted.

**Comment by JohnTitor** [comment]
Ping from triage: @Karrq and @GuillaumeGomez any updates on this?

**Comment by GuillaumeGomez** [comment]
I'm waiting on @Karrq. :)

**Comment by Karrq** [comment]
Doesn't look like that to me. I added 3 comments to the issue and a commit based on those comments. Haven't received a review/reply since.

I see that github says that there are changes requested, but I can't figure out how to remove that since the changed have happened.

@GuillaumeGomez @JohnTitor

**Review by Karrq — COMMENTED** [review]
Outdated

**Comment by Karrq** [comment]
So, @GuillaumeGomez, as I [commented](https://github.com/rust-lang/rust/pull/63022#issuecomment-515688912) [previously](https://github.com/rust-lang/rust/pull/63022#issuecomment-515698953) there are some concerns that I think need to be discussed. 

I'd like to know what you are waiting on me for. Is is the build errors for the pipeline? If that's so I'd like some guidance on how to resolve them, since I'm not able to figure out why they are being generated in the first place.

**Comment by GuillaumeGomez** [comment]
The error index aren't generated automatically, when you generate a new compiler error, you create a new error index alongside and normally you also provide the long error explanation (and a test where the error is checked to be emitted). It's possible that the path to the error isn't possible anymore and then you should just try your luck with another error code (one which **has** a test).

**Comment by hdhoang** [comment]
ping from triage @Karrq, any updates on this? Thanks for your time!

**Comment by Karrq** [comment]
Unfortunately I've been very busy recently and couldn't really take a look at the error index, but since the error itself doesn't seem to be able to be emitted I don't think there's much more I can do

**Comment by JohnCSimon** [comment]
From triage: 
Thank you for the update.
@Karrq @GuillaumeGomez

**Comment by GuillaumeGomez** [comment]
I'll need to check later in the future.

**Comment by JohnCSimon** [comment]
Ping from triage:
@Karrq @GuillaumeGomez Can you please resolve the failed checks?
Thank you.

**Comment by JohnTitor** [comment]
Ping from triage: @Karrq @GuillaumeGomez any updates on this?

**Comment by joelpalmer** [comment]
Ping from Triage: @Karrq we are closing this due to inactivity. However, when you have updates, please re-open. Thank you for the PR.

**Pull request closed without merging**